### DISCUSSION TITLE
Component Shadowing - Migrating from Gatsby

### DISCUSSION THREAD

**Discussion by styxlab** [vercel/next.js#17707] [Help]
I just read this excellent guide on [Migrating from Gatsby](https://nextjs.org/docs/migrating/from-gatsby). There is one Gatsby feature I heavily rely on: component shadowing with a plugin approach. Basically you can install plugins, that overwrite/shadow parts of the code which makes it super easy to extend and contribute to existing projects without touching the original code base.

I haven't seen anything on Next.js which could be used for component shadowing - but maybe there is a clever way to do this in Next.js as well? If so, let me know!

**Comment by lfades** [comment]
@leerob ☝️ Any thoughts here? 🙏

**Reply by leerob** [reply]
Given there isn't an established plugin system yet (still in RFC), this question doesn't have a great 1:1 with Next. There are ways to achieve the same _end result_, but its implementation isn't going to look anything like what Gatsby has. 

It would probably be more beneficial in this case to step back and forget about Gatsby and just ask what problem you're trying to solve.

**Reply by styxlab** [reply]
Thanks for your quick response! Could you please give a reference to the mentioned RFC so I can follow up on it? 

Here is the use case: I have been thinking about migrating [gatsby-starter-try-ghost](https://github.com/styxlab/gatsby-starter-try-ghost) to Next.js, but didn't find a convincing way to migrate the plugins. `gatsby-starter-try-ghost` relies on a plugin approach to give users the ability to easily add new features. Moreover, the project works well with some other third party Gatsby plugins (e.g. for site tracking). This is possible due to [component shadowing](https://www.gatsbyjs.com/docs/themes/shadowing/). It's also a great way for collaborators to add functionality with clear boundaries.

I understand that Next.js doesn't have a plugin system and I am open to other approaches too. So, if there is way to provide additional features through user plugins, I'd be very excited to look into that.

**Reply by wagerfield** [reply]
@leerob I too would love to get involved with the plugins RFC conversation.

When deciding between Next and Gatsby for a couple of pretty large client sites (~50K pages per locale) we ultimately decided to go with Gatsby due to the large ecosystem of plugins.

Having now built those large scale sites with Gatsby and learning the hard way that it _really doesn't scale_, I would like to bring my learnings and suggestions to the table.

**Reply by leerob** [reply]
@wagerfield Feel free to leave thoughts and ideas here https://github.com/vercel/next.js/discussions/9133

**Comment by ZeldOcarina** [comment]
This is very good to build extensible themes where you can basically adapt an existing website for many different customers with small shadowing on top of the existing theme on a monorepo..
